### ISSUE TITLE
Support OnApplicationStart[Async] method

### ISSUE THREAD

**Issue by divega** [aspnet/Hosting#373]
Applications often need to run code once at application startup per web server, e.g. to setup or initialize some external resource to be used by the application. There are two main issues with doing so in `Startup.Configure()`:

1. Lifetime of services is incorrect at this stage can be incorrect, e.g. scoped services will not be disposed and will remain in memory
2. It cannot be done asynchronously  

One general idea that has been discussed is to have a middleware that will run this code on first request but use a lock to make sure it is only done once.

The main example where this have come up is #3070 "Pattern for seeding database with EF7 and ASP.NET 5". Note that we cannot recommend a pattern like this in general because it is not safe unless you know there is only one web server, but for many customers this can be enough. Note that MusicStore [currently uses a workaround](https://github.com/aspnet/MusicStore/blob/dev/src/MusicStore/Models/SampleData.cs#L23) for the first issue.

Another possibly related scenario is described at https://github.com/aspnet/EntityFramework/issues/3070#issuecomment-140017643, although it is not about code that needs to run at application start, it is about how to get services to be activated with the right lifetime on code that doesn't execute as part of a request:

> Another problematic case is when you use a Timer. During the callback there have no scope. So the used instances are always the same, because you have to use the ServiceProvider of the application, so always the same context.

What would be the best way to use DI with a Timer? The Timer sould perhaps create a scope during the callback and pass it in parameter? Use IServiceScopeFactory in the callback? 

cc @Eilon

**Comment by Eilon** [comment]
I think we had a bug on this a million years ago but who knows where that bug is...

I've always liked the middleware approach to this. It's kind of like the IIS warmup module.

**Comment by PinpointTownes** [comment]
> I think we had a bug on this a million years ago but who knows where that bug is...

Just there, sir: https://github.com/aspnet/Hosting/issues/29 :smile:

**Comment by Eilon** [comment]
Yeah that's just about a million years ago! :smile:

**Comment by Eilon** [comment]
@Tratcher and recent thoughts on this?

**Comment by Tragetaschen** [comment]
What about [ApplicationStarted](https://github.com/aspnet/Hosting/blob/dev/src/Microsoft.AspNet.Hosting/ApplicationLifetime.cs#L22)?

**Comment by Eilon** [comment]
@Tragetaschen the scenario is an application warmup scenario, such as hydrating a cache, etc. So the requirement is to have some sort of "event" that is *blocking* and *async*. Not sure that cancellation token fits the bill.

**Comment by Tratcher** [comment]
I think this really belongs in Startup.Configure.
1. If there's an issue with scoped services in Configure then we should fix it. Making a scope for Configure shouldn't be a big deal.
2. Async at startup is meaningless if you need the app to block until you're done anyways.

The middleware approach is not good for app initialization because it is only triggered by the first request, and that request needs to block all other requests while it does it's work. When it's done that middleware stays in the pipeline and runs on every request.

**Comment by Eilon** [comment]
I don't know that the scopes can really be "fixed" because there's nothing actually wrong with it. The problem is that the app's "warmup" code requires those services, but is not running within the desired scope. The idea of a middleware (though perhaps it's not the right solution) is that the middleware would obviously run in the right scope (request scope).

I agree that actually being async is meaningless because it's blocking, but it allows the app developer to write non-ugly code (avoid calling `Wait` or `.Result` on everything).

As far as running on every request, well, the cost would presumably be negligible...

**Comment by gongdo** [comment]
Any updates about this?
Is [this](https://github.com/aspnet/MusicStore/blob/221bf3049219d66730d308f538f4a613cf2c9430/src/MusicStore/Startup.cs) a recommendation for async-startup(or warm-up)?

@Eilon I agree that the cost of initialization blocking would be negligible, however It doesn't seem to be nice if I should write a blocking middleware for one-time initialization.
I'd like to choose `Wait` or `Result` rather than blocking middleware.

**Comment by muratg** [comment]
@Tratcher @Eilon Are we doing this or backlog?

**Comment by Eilon** [comment]
No new features for now. Backlog.

**Comment by pauldotknopf** [comment]
Is this still visible and on the backlog? I notice the tag says "rc2".

**Comment by davidfowl** [comment]
No, there's no plan to do this.

**Comment by DenisDollfus** [comment]
We do some http request during startup to get all kind of configuration, so this feature definitely makes sense (no preference on the API shape specifically).

**Comment by chrispaynter** [comment]
This would be very helpful, especially when using platforms such as Contentful where in my case a lot of configuration lives. I need it at Startup but it takes an async network request to fetch it.

**Comment by Cyberboss** [comment]
I have a (albiet hacky) workaround for this using IApplicationLifetime.

```csharp
using Microsoft.AspNetCore.Builder;
using Microsoft.AspNetCore.Hosting;
using Microsoft.Extensions.DependencyInjection;
using System;
using System.Threading;
using System.Threading.Tasks;

namespace TGWebhooks.Core
{
    static class ApplicationBuilderExtensions
    {
		public static void UseAsyncInitialization<TInitializer>(this IApplicationBuilder app, Func<TInitializer, CancellationToken, Task> asyncInitializer)
		{
			var applicationLifetime = app.ApplicationServices.GetRequiredService<IApplicationLifetime>();
			Task initializationTask = null;
			applicationLifetime.ApplicationStarted.Register(() =>
			{
				var toInitialize = app.ApplicationServices.GetRequiredService<TInitializer>();
				initializationTask = asyncInitializer(toInitialize, applicationLifetime.ApplicationStopping);
			});

			app.Use(async (context, next) =>
			{
				await initializationTask;
				await next.Invoke();
			});
		}
	}
}
```
They must be the first handlers in the pipeline and block on the first request until finished. Allowing you to write code like this:
```csharp
public void Configure(IApplicationBuilder app, IHostingEnvironment env)
{
	app.UseAsyncInitialization<IPluginManager>((pluginManager, cancellationToken) => pluginManager.Initialize(cancellationToken));

	if (env.IsDevelopment())
                app.UseDeveloperExceptionPage();

        app.UseMvc();
}
```

**Comment by giggio** [comment]
The problem on async startup is still pending. Would you consider it for v3?

**Comment by dotnetjunkie** [comment]
I don't think there is any practical benefit in allowing start-up to be async. I like to refer to a [discussion about this](https://stackoverflow.com/questions/45924027/avoiding-all-di-antipatterns-for-types-requiring-asynchronous-initialization) I had with Stephen Cleary. My main point is:

> Typically, there is no real benefit of doing start-up initialization asynchronously. There is no practical performance benefit, because at start-up time, there will only be a single thread running anyway (although we might parallelize it, but that obviously doesn’t require async). Also note that although some application types might deadlock on doing synch-over-async, in the Composition Root we know exactly which application type we are using and whether or not this will be a problem or not. A Composition Root is application-specific. In other words, when we have initialization in our Composition Root, there is typically no benefit of doing start-up initialization asynchronously.

Although some application types might deadlock while blocking async calls, this is not the case in ASP.NET Core. There would therefore not be any benefit in adding all sorts of asynchronous start-up abstractions.

**Comment by Cyberboss** [comment]
Isn't this solved with [IHostedService](https://docs.microsoft.com/en-us/aspnet/core/fundamentals/host/hosted-services) now?

**Comment by Drawaes** [comment]
Two things

1. Using aspnet as a webserver inside some other application (it's small enough to allow this) would be useful if you could start it async.

2. It supports a cleaner model. We constantly tell users to never use result or wait and then say... Oh yeah but do it in startup which confuses the issue. The messaging should be never use wait or result and a single way of doing it. It's more a messaging problem in this case.

**Comment by giggio** [comment]
I agree with both of @Drawaes points. My main concern is number 2. There is definitely a message problem here, exactly as he states.

If you decide to not do it it would be a good idea to publish docs bringing up some best practices on how to do this correctly. I still have nightmares with deadlocks from `.Result` or `.Wait`. When I saw that there was no way to do this asynchronously the first thought that came to mind was uh-oh! 😱

**Comment by CoskunSunali** [comment]
Are there any known workarounds to this?

**Comment by matteocontrini** [comment]
@CoskunSunali I've been using [AspNetCore.AsyncInitialization](https://github.com/thomaslevesque/AspNetCore.AsyncInitialization) and it suits my needs. There's a [blog post](https://www.thomaslevesque.com/2018/09/25/asynchronous-initialization-in-asp-net-core-revisited/) from the author that explains it

**Comment by Eilon** [comment]
On one of my apps it was suggested to use an `async Program.Main` like so:

https://github.com/Eilon/Hubbup/blob/master/src/Hubbup.Web/Program.cs#L12-L24

This is from @Tratcher , so credit to him on this.

**Comment by Tratcher** [comment]
@CoskunSunali async Startup is not supported, and blocking is highly discouraged.

**Comment by abatishchev** [comment]
Just `Task.Run(() => ConfigurePluginServicesAsync()).Wait()` would also work and that's what everyone uses today, to my knowledge. And what this issue is all about. Even if async startup makes no sense methodologically.

**Comment by davidfowl** [comment]
That code that does a Task.Run followed by .Wait doesn’t make much sense. There’s no fear of deadlocking in ASP.NET Core applications so the additional Task.Run is kinda pointless.

We should look at async startup but today you can use an IHostedService to do you async initialization

**Comment by abatishchev** [comment]
The point is to be on the safe side of the equation, to always use same code on the sync-to-async boundary. Diversification of approaches in this case leads to mistakes. Any overhead is insignificant as the code runs just once.

**Comment by davidfowl** [comment]
I think it gives a far sense of security. It’s usually presented as “the right way to make async code synchronous”. I’ve seen all of the ways and there are all different tradeoffs. If I saw this code in an application I would guess it was copy pasta and the developer didn’t actually understand what was going on.
